### PR TITLE
Export of Openai Whisper with batched prompts

### PR DESCRIPTION
**Pull request by shubhambhokare1**
Adds an example to demonstrate the export of openai whipser implemenation with batch_size > 1 and addition of prompts for each audio snippet. 

Also handles the scenario for when prompts are not of the same size. For example if our prompt ids are [p1_id_1, p1_id_2] and [p2_id_1], the final decoder_input_ids will look as such after padding:
`[prev_token, p1_id_1, p1_id_2, start_token, lang_token, transcribe_token]
[prev_token, p2_id_1, PAD_TOKEN, start_token, lang_token, transcribe_token]`

### CODE REVIEW

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from pathlib import Path
 from typing import Dict, Tuple, Union
 
+import datasets
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:12]: Can you move the below logic here? This was added because some CI pipelines do not have `datasets` installed.

https://github.com/microsoft/onnxruntime/blob/7b46b3155891cbf6c783e2535e413b92ae81b050/onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py#L333-L342

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 2]
```diff
@@ -313,12 +314,103 @@ def optimize_onnx(
...
+
+        if prompt_mode:
+            prompts = ["John has doubts", "Maria has grave doubts"]
+            prompt_ids = [processor.get_prompt_ids(p) for p in prompts]
+            pt_transcription = []
+            pt_outputs = []
+            for i in range(batch_size):
+                inputs["prompt_ids"] = torch.from_numpy(prompt_ids[i])
+                inputs["input_features"] = input_features_[i].to(device)
+                pt_output = pt_model.generate(**inputs).detach().cpu().numpy()
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:359]: Can `pt_model.generate(**inputs)` be called directly with batch size > 1 or is the loop necessary?

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 3]
```diff
@@ -313,12 +314,103 @@ def optimize_onnx(
...
+        input_features_ = []
+        if batch_size == 1:
+            input_features = processor([ds[0]["audio"]["array"]], return_tensors="pt").input_features
+        else:
+            input_features_ = [
+                processor([ds[3]["audio"]["array"]], return_tensors="pt").input_features,
+                processor([ds[3]["audio"]["array"]], return_tensors="pt").input_features,
+            ]
+            assert len(input_features_) == batch_size
+            input_features = torch.cat((input_features_[0], input_features_[1])).to(device)
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:335]:
```suggestion
            input_features = torch.cat((input_features_[0], input_features_[1]))
```
The input features appear to be moved to the device in both cases.

https://github.com/microsoft/onnxruntime/blob/b34a04af6f79598ee8968c5ac0d82155c55de7ce/onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py#L339-L340

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 4]
```diff
@@ -398,36 +496,28 @@ def verify_onnx(
...
-        pt_transcription = processor.batch_decode(pt_outputs, skip_special_tokens=True)[0]
-        ort_transcription = processor.batch_decode(ort_outputs, skip_special_tokens=True)[0]
-
-        parity = (
-            pt_transcription in expected_transcription_options and ort_transcription in expected_transcription_options
-        )
+        ort_outputs = ort_session.run(None, inputs)[0]
+        ort_transcription = []
+        for o in ort_outputs:
+            ort_transcription.append(processor.batch_decode(o, skip_special_tokens=True)[0])
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:502]: Same question about whether the loop is necessary or not

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 5]
```diff
@@ -398,36 +496,28 @@ def verify_onnx(
...
-            else:
-                diff = pt_outputs - ort_outputs
-            max_diff = max(diff.min(), diff.max(), key=abs)
+            for i in range(batch_size):
+                if pt_outputs[i].shape != ort_outputs[i].shape:
+                    diff = pt_outputs[i] - ort_outputs[i][:, : len(pt_outputs[i])]
+                else:
+                    diff = pt_outputs[i] - ort_outputs[i]
+                max_diff_i = max(diff.min(), diff.max(), key=abs)
+            max_diff = max(max_diff, max_diff_i)
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:520]: I think this should be inside the loop to compare `max_diff` with each `max_diff_i`

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py (hunk 4 above)
Comment (shubhambhokare1) [reply onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:502]: Yup, i think we can get rid of the loop for this case

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py (hunk 2 above)
Comment (shubhambhokare1) [reply onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:359]: We cannot pass different prompts if we call it for batch_size > 1

Comment (shubhambhokare1) [reply onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:369]: We need to loop to pass different prompts for batch_size > 1

File: onnxruntime/python/tools/transformers/models/whisper/whisper_encoder_decoder_init.py [hunk 6]
```diff
@@ -132,8 +134,7 @@ def export_onnx(
         input_list = inputs.to_list()
 
         # TODO : Investigate whether copy of model if needed
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_encoder_decoder_init.py:136]:
```suggestion
```

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 7]
```diff
@@ -362,10 +357,97 @@ def verify_onnx(
...
         del inputs["early_stopping"]
         del inputs["use_cache"]
+        return inputs, pt_transcription, pt_outputs, prompt_ids
+
+    @staticmethod
+    def select_transcription_options(
+        batch_size: int,
+        prompt_mode: bool,
+    ):
+        if batch_size > 1 and prompt_mode is True:
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:391]:
```suggestion
        if batch_size > 1 and prompt_mode:
```

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 8]
```diff
@@ -362,10 +357,97 @@ def verify_onnx(
...
+        prompt_mode: bool = False,
+    ):
+        """Compare the result from PyTorch and ONNX Runtime to verify the ONNX model is good."""
+        extra_kwargs = {}
+        if version.parse(transformers_version) >= version.parse("4.36.0"):
+            extra_kwargs["attn_implementation"] = "eager"
+        pt_model = WhisperForConditionalGeneration.from_pretrained(
+            model_name_or_path, cache_dir=cache_dir, **extra_kwargs
+        ).to(device)
+        processor = WhisperProcessor.from_pretrained(model_name_or_path)
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:435]:
```suggestion
        processor = WhisperProcessor.from_pretrained(model_name_or_path, cache_dir=cache_dir)
```

File: onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py [hunk 9]
```diff
@@ -362,10 +357,97 @@ def verify_onnx(
...
+        config = WhisperConfig.from_pretrained(model_name_or_path)
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_helper.py:436]:
```suggestion
        config = WhisperConfig.from_pretrained(model_name_or_path, cache_dir=cache_dir)
```

File: onnxruntime/python/tools/transformers/models/whisper/whisper_openai_helper.py [hunk 10]
```diff
@@ -73,4 +75,10 @@ def forward(
         present_self = [
             present_val.reshape(present_val.shape[:2] + (-1, 64)).transpose(1, 2) for present_val in present_self
         ]
+
+        # Remove forward hooks to avoid model cloning step
+        if hooks is not None and remove_hooks is True:
```
Comment (kunal-vaishnavi) [inline onnxruntime/python/tools/transformers/models/whisper/whisper_openai_helper.py:80]:
```suggestion
        if hooks is not None and remove_hooks:
```

**Review by kunal-vaishnavi — APPROVED** [review]
